### DISCUSSION TITLE
Bi-directional relationship selectin loading with loader options issues not needed, inefficient queries

### DISCUSSION THREAD

**Discussion by igoras1993** [sqlalchemy/sqlalchemy#8663] [Usage Questions] [answered]
Hi, I am struggling with bi-directional relationships loading using `selectinload` at per-query basis. I am using most rececnt sqlalchemy release:
```
sqlalchemy==1.4.42
```

Consider following simple mapping

```python3
class Driver(Base):
    __tablename__ = "driver"

    id = Column(Integer, primary_key=True)
    name = Column(UnicodeText)

    plan_entries = relationship(
        "PlanEntry",
        back_populates="driver",
        lazy="raise",
        cascade="all, delete-orphan",
    )


class PlanEntry(Base):
    __tablename__ = "plan_entry"

    id = Column(Integer, primary_key=True)

    driver_id = Column(Integer, ForeignKey("driver.id"), nullable=False)
    driver = relationship(
        "Driver", back_populates="plan_entries", lazy="selectin"
    )
```
Note that so far `PlanEntry` has enabled `selectin` loading strategy at mapping time.

Assuming that I have populated two tables with one row each, now I want to issue the query against `Driver` with `PlanEntries` loaded with `selectin` strategy. Note the assert. What I am trying to achieve is to load `Driver` with its `.plan_entries`, and to be able to refer `.driver` property from this plan entry relationship.

```python3
with Session() as session:
    stmt = select(Driver).options(selectinload(Driver.plan_entries))
    d1 = session.execute(stmt).scalar_one()
    assert d1.plan_entries[0].driver == d1
```

It works as expected, with following queries issued to database:

```
2022-10-18 11:41:27,502 INFO sqlalchemy.engine.Engine BEGIN (implicit)
2022-10-18 11:41:27,504 INFO sqlalchemy.engine.Engine SELECT driver.id, driver.name 
FROM driver
2022-10-18 11:41:27,504 INFO sqlalchemy.engine.Engine [generated in 0.00014s] ()
2022-10-18 11:41:27,506 INFO sqlalchemy.engine.Engine SELECT plan_entry.driver_id AS plan_entry_driver_id, plan_entry.id AS plan_entry_id 
FROM plan_entry 
WHERE plan_entry.driver_id IN (?)
2022-10-18 11:41:27,506 INFO sqlalchemy.engine.Engine [generated in 0.00044s] (1,)
2022-10-18 11:41:27,507 INFO sqlalchemy.engine.Engine ROLLBACK
```

Now I will change mapping-time loading strategy to `raise` (desired default one) and define `selectin` at query time:

```python3
class PlanEntry(Base):
    # ...
    driver = relationship(
        "Driver", back_populates="plan_entries", lazy="raise"
    )

# ...

with Session() as session:
    stmt = select(Driver).options(
        selectinload(Driver.plan_entries).selectinload(PlanEntry.driver)
    )
    d1 = session.execute(stmt).scalar_one()
    assert d1.plan_entries[0].driver == d1

```

It also works, but this time **additional `select ... in` query component is issued**. It does not work the same as before. Look at the logs:

```
2022-10-18 11:50:25,258 INFO sqlalchemy.engine.Engine BEGIN (implicit)
2022-10-18 11:50:25,260 INFO sqlalchemy.engine.Engine SELECT driver.id, driver.name 
FROM driver
2022-10-18 11:50:25,260 INFO sqlalchemy.engine.Engine [generated in 0.00013s] ()
2022-10-18 11:50:25,263 INFO sqlalchemy.engine.Engine SELECT plan_entry.driver_id AS plan_entry_driver_id, plan_entry.id AS plan_entry_id 
FROM plan_entry 
WHERE plan_entry.driver_id IN (?)
2022-10-18 11:50:25,263 INFO sqlalchemy.engine.Engine [generated in 0.00032s] (1,)
2022-10-18 11:50:25,264 INFO sqlalchemy.engine.Engine SELECT driver.id AS driver_id, driver.name AS driver_name 
FROM driver 
WHERE driver.id IN (?)
2022-10-18 11:50:25,264 INFO sqlalchemy.engine.Engine [generated in 0.00019s] (1,)
2022-10-18 11:50:25,265 INFO sqlalchemy.engine.Engine ROLLBACK
```

I thought that this should be covered by `contains_eager` strategy, since the query indeed already contains everything needed to load `PlanEntry.driver` relationship. But using `contains_eager` produces even weirder behaviour:

```python3
with Session() as session:
    stmt = select(Driver).options(
        selectinload(Driver.plan_entries).contains_eager(PlanEntry.driver)
    )
    d1 = session.execute(stmt).scalar_one()
    assert d1.plan_entries[0].driver == d1

```
Now it produces cross product select:

```
2022-10-18 12:00:55,647 INFO sqlalchemy.engine.Engine BEGIN (implicit)
2022-10-18 12:00:55,648 INFO sqlalchemy.engine.Engine INSERT INTO driver (name) VALUES (?)
2022-10-18 12:00:55,648 INFO sqlalchemy.engine.Engine [generated in 0.00012s] ('Driver 1',)
2022-10-18 12:00:55,650 INFO sqlalchemy.engine.Engine INSERT INTO plan_entry (driver_id) VALUES (?)
2022-10-18 12:00:55,650 INFO sqlalchemy.engine.Engine [generated in 0.00015s] (1,)
2022-10-18 12:00:55,650 INFO sqlalchemy.engine.Engine COMMIT
2022-10-18 12:00:55,651 INFO sqlalchemy.engine.Engine BEGIN (implicit)
2022-10-18 12:00:55,652 INFO sqlalchemy.engine.Engine SELECT driver.id, driver.name 
FROM driver
2022-10-18 12:00:55,652 INFO sqlalchemy.engine.Engine [generated in 0.00012s] ()
/home/igor/code/sqla-rel-bi-eager/main.py:81: SAWarning: SELECT statement has a cartesian product between FROM element(s) "plan_entry" and FROM element "driver".  Apply join condition(s) between each element to resolve.
  d1 = session.execute(stmt).scalar_one()
2022-10-18 12:00:55,656 INFO sqlalchemy.engine.Engine SELECT plan_entry.driver_id AS plan_entry_driver_id, driver.id AS driver_id, driver.name AS driver_name, plan_entry.id AS plan_entry_id 
FROM plan_entry, driver 
WHERE plan_entry.driver_id IN (?)
2022-10-18 12:00:55,656 INFO sqlalchemy.engine.Engine [generated in 0.00049s] (1,)
2022-10-18 12:00:55,657 INFO sqlalchemy.engine.Engine ROLLBACK
```

**How to load bi-directional relationship using `selectin` strategy defined at per-query basis without inefficient queries?**

Here is the whole script for reproduction.

```python3
from sqlalchemy import (
    create_engine,
    select,
    Column,
    Integer,
    UnicodeText,
    ForeignKey,
)
from sqlalchemy.orm import (
    contains_eager,
    selectinload,
    registry,
    relationship,
    sessionmaker,
)
from sqlalchemy.pool import StaticPool

engine = create_engine("sqlite://", poolclass=StaticPool, echo=True)

Session = sessionmaker(
    autocommit=False,
    autoflush=False,
    expire_on_commit=False,
    future=True,
    bind=engine,
)

mapper_registry = registry()

Base = mapper_registry.generate_base()


class Driver(Base):
    __tablename__ = "driver"

    id = Column(Integer, primary_key=True)
    name = Column(UnicodeText)

    plan_entries = relationship(
        "PlanEntry",
        back_populates="driver",
        lazy="raise",
        cascade="all, delete-orphan",
    )


class PlanEntry(Base):
    __tablename__ = "plan_entry"

    id = Column(Integer, primary_key=True)

    driver_id = Column(Integer, ForeignKey("driver.id"), nullable=False)
    driver = relationship(
        "Driver", back_populates="plan_entries", lazy="raise"
    )


def initialize_data():

    mapper_registry.metadata.drop_all(bind=engine)
    mapper_registry.metadata.create_all(bind=engine)

    with Session() as session:

        d1 = Driver(name="Driver 1")
        pe1 = PlanEntry(driver=d1)

        session.add(d1)
        session.commit()


if __name__ == "__main__":
    initialize_data()

    with Session() as session:

        stmt = select(Driver).options(
            selectinload(Driver.plan_entries).contains_eager(PlanEntry.driver)
        )

        d1 = session.execute(stmt).scalar_one()

        assert d1.plan_entries[0].driver == d1

```

**Comment by igoras1993** [comment]
Some update here. I've struggled a little bit more with this and found out that doing:

```
stmt = select(Driver).options(
    selectinload(Driver.plan_entries).lazyload(PlanEntry.driver)
)
```
Does the job. But I am afraid that this is actually a [perfectly fine] effect of session being an [identity map](https://docs.sqlalchemy.org/en/14/glossary.html#term-identity-map), **not an actual answer for my concerns**. 

What I would really prefer is this:
```
stmt = select(Driver).options(
    selectinload(Driver.plan_entries)
)
```

Since `Driver` is already there, and `contains_eager` seems to behave like it was designed to work only with explicit JOINs (is this the case btw?), I am assuming that the only thing that blocks me from such usage is actual `raise` load sitting in the model as default strategy. Setting strategy to `lazyload` at per-query basis doesn't do any magic, just disables the `raiseload` exception raising bahaviour. Am I correct?

Since above, I would argue that the most obvious way to disable default `raiseload` (enabling `back_poulating` for `PlanEntry.driver`)  would be the `contains_eager` method. But I have shown that this introduces (very suprisingly) a cartesian product. Am I missing something?

**Comment by zzzeek** [comment, marked as answer]
> 
> Now I will change mapping-time loading strategy to `raise` (desired default one) and define `selectin` at query time:
> 
> ```python
> class PlanEntry(Base):
>     # ...
>     driver = relationship(
>         "Driver", back_populates="plan_entries", lazy="raise"
>     )
> 
> # ...
> 
> with Session() as session:
>     stmt = select(Driver).options(
>         selectinload(Driver.plan_entries).selectinload(PlanEntry.driver)
>     )
>     d1 = session.execute(stmt).scalar_one()
>     assert d1.plan_entries[0].driver == d1
> ```
> 
> It also works, but this time **additional `select ... in` query component is issued**. It does not work the same as before. Look at the logs:

that's correct because you told it to load PlanEntry.driver using selectinload().

> I thought that this should be covered by `contains_eager` strategy, since the query indeed already contains everything needed to load `PlanEntry.driver` relationship. But using `contains_eager` produces even weirder behaviour:
> 
> ```python
> with Session() as session:
>     stmt = select(Driver).options(
>         selectinload(Driver.plan_entries).contains_eager(PlanEntry.driver)
>     )
>     d1 = session.execute(stmt).scalar_one()
>     assert d1.plan_entries[0].driver == d1
> ```

you can't use contains_eager() against selectinload() as it makes no sense.  I've never seen anyone try that but that should be changed to raise an error.

> **How to load bi-directional relationship using `selectin` strategy defined at per-query basis without inefficient queries?**

I'm not sure what the goal is here; the original mapping loaded both sides of the bi-directional relationship with just two queries.   Adding an additional selectinload() for a many-to-one is unnecessary.   I would not use selectinload() for many-to-one in any case.

For simple many-to-one relationships, the ORM will pull that `Driver` object from the identity map without having to emit any SQL.      For this reason, when you want to use "lazy='raise'" on a many-to-one, you might instead choose to use ``lazy="raise_on_sql"``, so that it can still "lazy load" from the local identity map, it would only raise if it actually had to emit a SQL query.

**Reply by igoras1993** [reply]
> For this reason, when you want to use "lazy='raise'" on a many-to-one, you might instead choose to use lazy="raise_on_sql", so that it can still "lazy load" from the local identity map, it would only raise if it actually had to emit a SQL query.

This is exactly what I was looking for. Thank You very much!
